File: java/std/src/bench/Skynet.java
// The skynet benchmark as described here:
// https://github.com/atemerev/skynet
//
// Canonical (TooManyCooks) implementation:
// ../../../../cpp/TooManyCooks/skynet.cpp
//
// Each node forks 10 children until DEPTH_MAX levels deep (10^DEPTH_MAX = 100M
// leaf tasks) and sums their results. Here each child is a virtual thread, so
// this measures virtual-thread spawn/join throughput under deep nested fan-out.
package bench;

public final class Skynet {
  private static final int DEPTH_MAX = 8;
  private static final int ITER_COUNT = 1;

  static long skynetOne(long base, long depth) {
    if (depth == DEPTH_MAX) {
      return base;
    }
    long depthOffset = 1;
    for (long i = 0; i < DEPTH_MAX - depth - 1; i++) {
      depthOffset *= 10;
    }
    final long offset = depthOffset; // effectively final for capture below

    // Each child writes its own slot, so the disjoint writes need no
    // synchronization beyond join()'s happens-before edge.
    long[] results = new long[10];
    Thread[] threads = new Thread[10];
    for (int i = 0; i < 10; i++) {
      final int idx = i;
      threads[i] = Thread.startVirtualThread(
        () -> results[idx] = skynetOne(base + offset * idx, depth + 1)
      );
    }
    Bench.joinAll(threads);

    long count = 0;
    for (long r : results) {
      count += r;
    }
    return count;
  }

  static void skynet(long expected) {
    long count = skynetOne(0, 0);
    if (count != expected) {
      System.out.printf("ERROR: wrong result - %d%n", count);
    }
  }

  public static void main(String[] args) {
    int threadCount = Bench.configureThreads(args, 0);

    long leaves = 1;
    for (int i = 0; i < DEPTH_MAX; i++) {
      leaves *= 10;
    }
    long expected = (leaves - 1) * leaves / 2;

    System.out.println("threads: " + threadCount);

    skynet(expected); // warmup

    System.out.println("runs:");
    long start = System.nanoTime();
    for (int i = 0; i < ITER_COUNT; i++) {
      skynet(expected);
    }
    long durUs = (System.nanoTime() - start) / 1000;
    System.out.println("  - iteration_count: " + ITER_COUNT);
    System.out.println("    duration: " + durUs + " us");
    System.out.println("    max_rss: " + Bench.peakMemoryUsageKiB() + " KiB");
  }
}

File: kotlin/fjp/skynet.kt
// The skynet benchmark as described here:
// https://github.com/atemerev/skynet
//
// Canonical (TooManyCooks) implementation:
// ../../cpp/TooManyCooks/skynet.cpp
//
// Each node forks 10 children until depthMax levels deep (10^depthMax = 100M
// leaf tasks) and sums their results. Here each child is a coroutine started via
// async {}, so this measures Kotlin's coroutine spawn/join throughput under deep
// nested fan-out. The scheduler under test is selected by benchutil.newDispatcher.
@file:JvmName("Main")

package skynet

import benchutil.newDispatcher
import benchutil.peakMemoryUsageKiB
import benchutil.threadCountArg
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking

private const val DEPTH_MAX = 8L
private const val ITER_COUNT = 1

private suspend fun skynetOne(base: Long, depth: Long): Long = coroutineScope {
    if (depth == DEPTH_MAX) {
        return@coroutineScope base
    }
    var offset = 1L
    repeat((DEPTH_MAX - depth - 1).toInt()) { offset *= 10 }

    // Each child computes an independent slot; coroutineScope's join gives the
    // happens-before needed to read them all back below.
    (0L until 10L)
        .map { i -> async { skynetOne(base + offset * i, depth + 1) } }
        .sumOf { it.await() }
}

private suspend fun skynet(expected: Long) {
    val count = skynetOne(0, 0)
    if (count != expected) {
        println("ERROR: wrong result - $count")
    }
}

fun main(args: Array<String>) {
    val threadCount = threadCountArg(args, 0)

    var leaves = 1L
    repeat(DEPTH_MAX.toInt()) { leaves *= 10 }
    val expected = (leaves - 1) * leaves / 2

    println("threads: $threadCount")

    runBlocking(newDispatcher(threadCount)) {
        skynet(expected) // warmup

        println("runs:")
        val start = System.nanoTime()
        repeat(ITER_COUNT) { skynet(expected) }
        val durUs = (System.nanoTime() - start) / 1000
        println("  - iteration_count: $ITER_COUNT")
        println("    duration: $durUs us")
        println("    max_rss: ${peakMemoryUsageKiB()} KiB")
    }
}
